# --- T2-COPYRIGHT-NOTE-BEGIN ---
# This copyright note is auto-generated by ./scripts/Create-CopyPatch.
# 
# T2 SDE: package/.../dietlibc/libm.patch
# Copyright (C) 2006 The T2 SDE Project
# 
# More information can be found in the files COPYING and README.
# 
# This patch file is dual-licensed. It is available under the license the
# patched project is licensed under, as long as it is an OpenSource license
# as defined at http://www.opensource.org/ (e.g. BSD, X11) or under the terms
# of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
# --- T2-COPYRIGHT-NOTE-END ---

This patch let's the ARM build continue, as well as let gawk build on ppc.
However, these math functions stubs have to be replaced with real generic
functions soon.
  - Juergen Sawinski
Original comment:
  - Rene Rebe <rene@exactcode.de>


diff -NurpP --minimal dietlibc-0.29-vanilla/arm/Makefile.add dietlibc-0.29/arm/Makefile.add
--- dietlibc-0.29-vanilla/arm/Makefile.add	2002-05-09 03:05:10.000000000 +0200
+++ dietlibc-0.29/arm/Makefile.add	2006-03-31 11:10:36.910869300 +0200
@@ -1,5 +1,6 @@
 
 LIBOBJ+=$(OBJDIR)/md5asm.o
+LIBMATH=$(patsubst libm/%.c,%.o,$(wildcard libm/generic-*.c))
 CFLAGS+=-Os -fomit-frame-pointer -fstrict-aliasing -mhard-float
 VPATH:=arm:syscalls.s:$(VPATH)
 LIBGMON_OBJS+=$(OBJDIR)/mcount.o
diff -NurpP --minimal dietlibc-0.29-vanilla/ppc/Makefile.add dietlibc-0.29/ppc/Makefile.add
--- dietlibc-0.29-vanilla/ppc/Makefile.add	2001-01-30 16:01:20.000000000 +0100
+++ dietlibc-0.29/ppc/Makefile.add	2006-03-31 11:11:46.986591030 +0200
@@ -1,3 +1,4 @@
 
+LIBMATH+= generic-atan2.o generic-ceil.o
 CFLAGS+=-mpowerpc-gpopt -mpowerpc-gfxopt -Os
 VPATH:=ppc:syscalls.s:$(VPATH)
diff -NurpP --minimal dietlibc-0.29-vanilla/ppc64/Makefile.add dietlibc-0.29/ppc64/Makefile.add
--- dietlibc-0.29-vanilla/ppc64/Makefile.add	2003-10-10 15:37:34.000000000 +0200
+++ dietlibc-0.29/ppc64/Makefile.add	2006-03-31 11:12:00.394241884 +0200
@@ -1,3 +1,4 @@
 
+LIBMATH+= generic-atan2.o generic-ceil.o
 CFLAGS+=-Os -mpowerpc64
 VPATH:=ppc64:syscalls.s:$(VPATH)

diff -NurpP --minimal dietlibc-0.29-vanilla/Makefile dietlibc-0.29/Makefile
--- dietlibc-0.29-vanilla/Makefile	2005-05-18 16:38:55.000000000 +0200
+++ dietlibc-0.29/Makefile	2006-03-31 11:04:21.656629687 +0200
@@ -112,7 +112,7 @@ LIBCRUFTOBJ=$(patsubst libcruft/%.c,$(OB
 LIBCRYPTOBJ=$(patsubst libcrypt/%.c,$(OBJDIR)/%.o,$(wildcard libcrypt/*.c))
 LIBSHELLOBJ=$(patsubst libshell/%.c,$(OBJDIR)/%.o,$(wildcard libshell/*.c))
 LIBCOMPATOBJ=$(patsubst libcompat/%.c,$(OBJDIR)/%.o,$(wildcard libcompat/*.c)) $(OBJDIR)/syscall.o
-LIBMATH=$(patsubst libm/%.c,%.o,$(wildcard libm/*.c))
+LIBMATH=$(patsubst libm/%.c,%.o,$(filter-out libm/generic-%,$(wildcard libm/*.c)))
 
 LIBRPCOBJ=$(patsubst librpc/%.c,$(OBJDIR)/%.o,$(wildcard librpc/*.c))
 LIBREGEXOBJ=$(patsubst libregex/%.c,$(OBJDIR)/%.o,$(wildcard libregex/*.c))
diff -NurpP --minimal dietlibc-0.29-vanilla/libm/generic-__half.c dietlibc-0.29/libm/generic-__half.c
--- dietlibc-0.29-vanilla/libm/generic-__half.c	1970-01-01 01:00:00.000000000 +0100
+++ dietlibc-0.29/libm/generic-__half.c	2006-03-31 11:14:27.631445932 +0200
@@ -0,0 +1,4 @@
+
+float __half = 0.5;
+
+
diff -NurpP --minimal dietlibc-0.29-vanilla/libm/generic-atan2.c dietlibc-0.29/libm/generic-atan2.c
--- dietlibc-0.29-vanilla/libm/generic-atan2.c	1970-01-01 01:00:00.000000000 +0100
+++ dietlibc-0.29/libm/generic-atan2.c	2006-03-31 11:02:54.908834172 +0200
@@ -0,0 +1,4 @@
+#include "libm/stub.h"
+#include <stdlib.h>
+#include <math.h>
+LIBM_STUB2(atan2)
diff -NurpP --minimal dietlibc-0.29-vanilla/libm/generic-ceil.c dietlibc-0.29/libm/generic-ceil.c
--- dietlibc-0.29-vanilla/libm/generic-ceil.c	1970-01-01 01:00:00.000000000 +0100
+++ dietlibc-0.29/libm/generic-ceil.c	2006-03-31 11:02:45.332512696 +0200
@@ -0,0 +1,4 @@
+#include "libm/stub.h"
+#include <stdlib.h>
+#include <math.h>
+LIBM_STUB1(ceil)
diff -NurpP --minimal dietlibc-0.29-vanilla/libm/generic-cos.c dietlibc-0.29/libm/generic-cos.c
--- dietlibc-0.29-vanilla/libm/generic-cos.c	1970-01-01 01:00:00.000000000 +0100
+++ dietlibc-0.29/libm/generic-cos.c	2006-03-31 11:02:45.333512521 +0200
@@ -0,0 +1,4 @@
+#include "libm/stub.h"
+#include <stdlib.h>
+#include <math.h>
+LIBM_STUB1(cos)
diff -NurpP --minimal dietlibc-0.29-vanilla/libm/generic-exp.c dietlibc-0.29/libm/generic-exp.c
--- dietlibc-0.29-vanilla/libm/generic-exp.c	1970-01-01 01:00:00.000000000 +0100
+++ dietlibc-0.29/libm/generic-exp.c	2006-03-31 11:02:45.333512521 +0200
@@ -0,0 +1,4 @@
+#include "libm/stub.h"
+#include <stdlib.h>
+#include <math.h>
+LIBM_STUB1(exp)
diff -NurpP --minimal dietlibc-0.29-vanilla/libm/generic-floor.c dietlibc-0.29/libm/generic-floor.c
--- dietlibc-0.29-vanilla/libm/generic-floor.c	1970-01-01 01:00:00.000000000 +0100
+++ dietlibc-0.29/libm/generic-floor.c	2006-03-31 11:02:45.333512521 +0200
@@ -0,0 +1,4 @@
+#include "libm/stub.h"
+#include <stdlib.h>
+#include <math.h>
+LIBM_STUB1(floor)
diff -NurpP --minimal dietlibc-0.29-vanilla/libm/generic-fmod.c dietlibc-0.29/libm/generic-fmod.c
--- dietlibc-0.29-vanilla/libm/generic-fmod.c	1970-01-01 01:00:00.000000000 +0100
+++ dietlibc-0.29/libm/generic-fmod.c	2006-03-31 11:03:06.734761356 +0200
@@ -0,0 +1,4 @@
+#include "libm/stub.h"
+#include <stdlib.h>
+#include <math.h>
+LIBM_STUB2(fmod)
diff -NurpP --minimal dietlibc-0.29-vanilla/libm/generic-log.c dietlibc-0.29/libm/generic-log.c
--- dietlibc-0.29-vanilla/libm/generic-log.c	1970-01-01 01:00:00.000000000 +0100
+++ dietlibc-0.29/libm/generic-log.c	2006-03-31 11:02:45.333512521 +0200
@@ -0,0 +1,4 @@
+#include "libm/stub.h"
+#include <stdlib.h>
+#include <math.h>
+LIBM_STUB1(log)
diff -NurpP --minimal dietlibc-0.29-vanilla/libm/generic-sin.c dietlibc-0.29/libm/generic-sin.c
--- dietlibc-0.29-vanilla/libm/generic-sin.c	1970-01-01 01:00:00.000000000 +0100
+++ dietlibc-0.29/libm/generic-sin.c	2006-03-31 11:02:45.334512346 +0200
@@ -0,0 +1,4 @@
+#include "libm/stub.h"
+#include <stdlib.h>
+#include <math.h>
+LIBM_STUB1(sin)
diff -NurpP --minimal dietlibc-0.29-vanilla/libm/generic-sqrt.c dietlibc-0.29/libm/generic-sqrt.c
--- dietlibc-0.29-vanilla/libm/generic-sqrt.c	1970-01-01 01:00:00.000000000 +0100
+++ dietlibc-0.29/libm/generic-sqrt.c	2006-03-31 11:02:45.334512346 +0200
@@ -0,0 +1,4 @@
+#include "libm/stub.h"
+#include <stdlib.h>
+#include <math.h>
+LIBM_STUB1(sqrt)
diff -NurpP --minimal dietlibc-0.29-vanilla/libm/generic-tan.c dietlibc-0.29/libm/generic-tan.c
--- dietlibc-0.29-vanilla/libm/generic-tan.c	1970-01-01 01:00:00.000000000 +0100
+++ dietlibc-0.29/libm/generic-tan.c	2006-03-31 11:02:45.334512346 +0200
@@ -0,0 +1,4 @@
+#include "libm/stub.h"
+#include <stdlib.h>
+#include <math.h>
+LIBM_STUB1(tan)
diff -NurpP --minimal dietlibc-0.29-vanilla/libm/stub.h dietlibc-0.29/libm/stub.h
--- dietlibc-0.29-vanilla/libm/stub.h	1970-01-01 01:00:00.000000000 +0100
+++ dietlibc-0.29/libm/stub.h	2006-03-31 11:09:39.534922846 +0200
@@ -0,0 +1,40 @@
+#ifndef _LIBM_STUB_H
+#define _LIBM_STUB_H
+
+#define _LIBM_STUB0(type,func) \
+  type func() { \
+    _write2("Function stub: " #func " (not implemented)\n"); \
+	abort(); \
+  }
+
+#define _LIBM_STUB1(type,func) \
+  type func(type a) { \
+    _write2("Function stub: " #func " (not implemented)\n"); \
+    abort(); \
+  }
+
+#define _LIBM_STUB2(type,func) \
+  type func(type a, type b) { \
+    _write2("Function stub: " #func " (not implemented)\n"); \
+    abort(); \
+  }
+
+#define _LIBM_STUB3(type,func) \
+  type func(type a, type b, type c) { \
+    _write2("Function stub: " #func " (not implemented)\n"); \
+    abort(); \
+  }
+
+#define LIBM_STUB0(func) \
+  _LIBM_STUB0(double, func)
+
+#define LIBM_STUB1(func) \
+  _LIBM_STUB1(double, func)
+
+#define LIBM_STUB2(func) \
+  _LIBM_STUB2(double, func)
+
+#define LIBM_STUB3(func) \
+  _LIBM_STUB3(double, func)
+
+#endif
